ddtrace/contrib/rq/__init__.py

```diff
@@ -0,0 +1,245 @@
+
+   If ``True`` the integration will connect the traces sent between the queuer
+   and the RQ worker.
+
+   Default: ``False``
+
+.. py:data:: ddtrace.config.rq['service']
+
+   The service name reported by default for RQ spans from the app.
+
+   This option can also be set with the ``DD_RQ_SERVICE`` or ``DD_SERVICE``
+   environment variables.
+
+   Default: ``rq``
+
+.. py:data:: ddtrace.config.rq_worker['service']
+
+   The service name reported by default for RQ spans from workers.
+
+   This option can also be set with the ``DD_RQ_WORKER_SERVICE`` environment
+   variable.
+
+   Default: ``rq-worker``
+"""
+from ddtrace import Pin
+from ddtrace import config
+
+from .. import trace_utils
+from ...ext import SpanTypes
+from ...propagation.http import HTTPPropagator
+from ...utils import get_argument_value
+
+
+__all__ = [
+    "patch",
+    "unpatch",
+]
+
+
+config._add(
+    "rq",
+    dict(
+        _default_service="rq",
+        distributed_tracing_enabled=False,
+    ),
+)
+
+config._add(
+    "rq_worker",
+    dict(
+        _default_service="rq-worker",
+    ),
+)
+
+
+@trace_utils.with_traced_module
+def traced_queue_enqueue_job(rq, pin, func, instance, args, kwargs):
+    job = get_argument_value(args, kwargs, 0, "f")
+
+    with pin.tracer.trace(
+        "rq.queue.enqueue_job",
+        service=trace_utils.int_service(pin, config.rq),
+        span_type=SpanTypes.WORKER,
+        resource=job.func_name,
+    ) as span:
+        span._set_str_tag("queue.name", instance.name)
+        span.set_tag("job.id", job.get_id())
+        span._set_str_tag("job.func_name", job.func_name)
+
+        # If the queue is_async then add distributed tracing headers to the job
+        if instance.is_async and config.rq.distributed_tracing_enabled:
+            HTTPPropagator.inject(span.context, job.meta)
+
+        try:
+            return func(*args, **kwargs)
+        finally:
+            # If the queue is not async then the job is run immediately so we can
+            # report the results
+            if not instance.is_async:
+                span.set_tag("job.status", job.get_status())
+
+
+@trace_utils.with_traced_module
+def traced_queue_fetch_job(rq, pin, func, instance, args, kwargs):
+    with pin.tracer.trace("rq.queue.fetch_job", service=trace_utils.int_service(pin, config.rq)) as span:
+        job_id = get_argument_value(args, kwargs, 0, "job_id")
+        span.set_tag("job.id", job_id)
+        return func(*args, **kwargs)
+
+
+@trace_utils.with_traced_module
+def traced_perform_job(rq, pin, func, instance, args, kwargs):
+    """Trace rq.Worker.perform_job"""
+    # `perform_job` is executed in a freshly forked, short-lived instance
+    job = get_argument_value(args, kwargs, 0, "job")
+
+    ctx = HTTPPropagator.extract(job.meta)
+    if ctx.trace_id:
+        pin.tracer.context_provider.activate(ctx)
+
+    try:
+        with pin.tracer.trace(
+            "rq.worker.perform_job",
```

Comment:
This operation name seems to be redundant if we also have a service name for the worker that is different from the enqueueing app.

It seems to me that we could reduce this by using `rq.queue` and `rq.worker` for the service names and remove them from the namespace of the operation names.

The job perform and fetch would then be under the `rq.worker` (though maybe I'm getting the flow wrong).

Reply:
however, this doesn't account for people having their own custom service names, keeping `rq.` in the operation name is good unless we want to start using `component` (or equivalent) tag.

Reply:
it also more closely matches the rq api which helps I think from a user perspective knowing what this span is wrapping